runtime/interpreter/value.go

```diff
@@ -18514,11 +18572,8 @@ func (v *DictionaryValue) iterateKeys(
 			panic(errors.NewExternalError(err))
 		}
 	}
-	if v.IsResourceKinded(interpreter) {
-		interpreter.withMutationPrevention(v.StorageID(), iterate)
-	} else {
-		iterate()
-	}
+
+	interpreter.withMutationPrevention(v.StorageID(), iterate)
 }
```

Comment:
call `interpreter.withMutationPrevention` for all values (both resource-typed and non-resource typed values)